lib/boot-module-files.js

```diff
@@ -0,0 +1,119 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2020 Joyent, Inc.
+ */
+
+/*
+ * Read / write boot-time module files
+ */
+
+const fs = require('fs');
+const path = require('path');
+const util = require('util');
+
+const assert = require('assert-plus');
+const mkdirp = require('mkdirp');
+const vasync = require('vasync');
+
+function generateBootModuleFile(opts, cb) {
+    assert.object(opts, 'opts');
+    assert.object(opts.log, 'opts.log');
+    assert.object(opts.module, 'opts.module');
+    assert.string(opts.module.content, 'opts.module.content');
+    assert.optionalString(opts.module.type, 'opts.module.type');
+
+    if (opts.module.type !== 'base64') {
+        cb(new Error(util.format(
+            'Unsupported type %s for module %s',
+            opts.module.type, opts.module.path)));
+        return;
+    }
+
+    cb(null, Buffer.from(
+        opts.module.content, 'base64').toString('ascii'));
```

Comment:
Suppose a module was some non-ascii blob, such as a gzipped file.  Does this work?

Reply:
No, it doesn't. This actually should probably be `utf8` instead of `ascii` though. The thing is that we need to provide an encoding for the `buffer.toString()` method or it will default to `utf8`.

If we want to allow passing specific encodings, we should limit to the list here https://nodejs.org/docs/latest-v12.x/api/buffer.html#buffer_buffers_and_character_encodings and probably figure out a way to add that information to the boot module, so we can use it when decoding the `base64` contents before writing the file.

Thoughts?

Reply:
I thought the point of using base64 encoding during transport was so that we could have arbitrary data in it.  I worry that decoding gzip data as utf8 could mangle it.  Why not binary or latin1?  Surely when these files are loaded via tftp, neither the tftp server nor anything on the client is going to treat it as anything other than a binary blob.

Reply:
Let's go for latin1 then.